sklearn/utils/estimator_checks.py

```diff
@@ -1732,6 +1732,29 @@ def check_transformer_preserve_dtypes(name, transformer_orig):
         )
 
 
+def check_regressor_preserve_dtypes(name, regressor_orig):
+    # Regressors predictions must have the same dtype
+    # than the one of the data.
+    regressor_orig = regressor_orig()
+
+    X, y = _regression_dataset()
+    X = _pairwise_estimator_convert_X(X, regressor_orig)
+    regressor = clone(regressor_orig)
+    y = _enforce_estimator_tags_y(regressor, y)
+
+    for dtype in _safe_tags(regressor_orig, key="preserves_dtype"):
+        y_cast = y.astype(dtype)
+        regressor = clone(regressor_orig)
+        regressor.fit(X, y_cast)
+        y_pred = regressor.predict(X)
```

Comment:
I think we should only check that `y_pred.dtype == dtype` if both `X` and `y` also share the same `dtype`:

```suggestion
        X_cast = X.astype(dtype)
        y_cast = y.astype(dtype)
        regressor = clone(regressor_orig)
        regressor.fit(X_cast, y_cast)
        y_pred = regressor.predict(X_cast)
```

other cases, are not expected to return values of specific dtype in general

Reply:
Thanks @ogrisel .